### PR TITLE
Buffer allocation pool for parsing protobuf

### PR DESCRIPTION
**Pull request by cevian**
Reduces allocations significantly.

### CODE REVIEW

File: pkg/api/parser/protobuf/protobuf.go [hunk 1]
```diff
@@ -12,14 +13,24 @@ import (
...
 		return fmt.Errorf("request body read error: %w", err)
 	}
 
-	if err = proto.Unmarshal(reqBuf, wr); err != nil {
+	if err = proto.Unmarshal(b.Bytes(), wr); err != nil {
 		return fmt.Errorf("protobuf unmarshal error: %w", err)
 	}
 
-	return nil
+	bufPool.Put(b)
```
Comment (Harkishen-Singh) [inline pkg/api/parser/protobuf/protobuf.go:28]: I think this should be in `defer` (line 17) as if we error out in line `21` or `25`, we do not put in the pool and this gets GCed.

File: pkg/api/parser/protobuf/protobuf.go [hunk 2]
```diff
@@ -12,14 +13,24 @@ import (
 // ParseRequest is responsible for populating the write request from the
 // data in the request in protobuf format.
 func ParseRequest(r *http.Request, wr *prompb.WriteRequest) error {
-	reqBuf, err := ioutil.ReadAll(r.Body)
+	b := bufPool.Get().(*bytes.Buffer)
+	b.Reset()
+
+	_, err := b.ReadFrom(r.Body)
```
Comment (Harkishen-Singh) [inline pkg/api/parser/protobuf/protobuf.go:19]: Can we do the same for read requests? That will help us save extra allocations.

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]

**Review by antekresic — APPROVED** [review]
I agree with @Harkishen-Singh on the defer change.

Other than that, looks good! :+1:

File: pkg/api/parser/protobuf/protobuf.go (hunk 2 above)
Comment (cevian) [reply pkg/api/parser/protobuf/protobuf.go]: Yes but in a separate PR. We should get flamegraphs to see we need this before doing it

File: pkg/api/parser/protobuf/protobuf.go (hunk 1 above)
Comment (cevian) [reply pkg/api/parser/protobuf/protobuf.go]: good catch

File: pkg/api/parser/protobuf/protobuf.go (hunk 2 above)
Comment (Harkishen-Singh) [reply pkg/api/parser/protobuf/protobuf.go]: @cevian TODO.

**Review by Harkishen-Singh — APPROVED** [review]
👍🏻

**Comment by Harkishen-Singh** [comment]
Please note the TODO in the comment. Maybe, let's have that on the project board otherwise.

**Pull request merged**